File: bot/helpers/tts.py
import os
import uuid
import logging
from bot.helpers.bash import bash
from bot.utils.piper import get_model

c = os.path.dirname(os.path.abspath(__file__))
models = os.path.join(c, "../models")
tmp = os.path.join(c, "../tmp")


async def tts(text: str, output_file: str):
    model = get_model()
    try:
        if not os.path.exists(models):
            os.makedirs(models)
        elif not os.path.exists(os.path.dirname(output_file)):
            os.makedirs(os.path.dirname(output_file))
        elif not os.path.exists(tmp):
            os.makedirs(tmp)
        tmpf = f"{tmp}/{uuid.uuid4()}.txt"
        with open(tmpf, "w") as temp_text:
            temp_text.write(text)
            temp_text.close()
        await bash(
            f"piper -m {model} --download-dir {models} --data-dir {models} -f {output_file} < {tmpf}"
        )
        os.remove(tmpf)
    except Exception as e:
        logging.error(e)
        raise e

File: kreacher/helpers/tts.go
package helpers

import (
	"fmt"
	"io"
	"os"
	"path"

	uuid "github.com/google/uuid"
	ini "gopkg.in/ini.v1"
)

func TTS(text string, opts ...string) (string, error) {
	tempFile, err := os.Create(fmt.Sprintf("%s.txt", uuid.New()))

	if err != nil {
		return "", err
	}

	_, err = io.WriteString(tempFile, text)

	if err != nil {
		return "", err
	}

	err = tempFile.Sync()

	if err != nil {
		return "", err
	}

	cfg, err := ini.Load("")

	if err != nil {
		return "", err
	}

	ast := cfg.Section("ADVANCED")

	piperPathKey, err := ast.GetKey("PIPER_DATA_PATH")

	if err != nil {
		return "", err
	}

	tempPathKey, err := ast.GetKey("TEMP_PATH")

	if err != nil {
		return "", err
	}

	piperModelsPath := path.Join(piperPathKey.String(), "models")

	_, err = os.Stat(piperModelsPath)

	if os.IsNotExist(err) {
		os.MkdirAll(piperModelsPath, 0755)
	}

	outputFile := path.Join(tempPathKey.String(), "tts", fmt.Sprintf("%s.wav", uuid.New()))

	_, err = os.Stat(path.Base(outputFile))

	if os.IsNotExist(err) {
		os.MkdirAll(outputFile, 0755)
	}

	stdout, err := Bash(fmt.Sprintf(
		"piper -m %s --download-dir %s --data-dir %s -f %s < %s",
		"",
		piperModelsPath,
		piperModelsPath,
		outputFile,
		tempFile.Name(),
	))

	if err != nil {
		return "", err
	}

	err = os.Remove(tempFile.Name())

	if err != nil {
		return "", err
	}

	return stdout, nil

}
